crates/language/src/language_registry.rs

```diff
@@ -509,20 +509,29 @@ impl LanguageRegistry {
         user_file_types: Option<&HashMap<Arc<str>, Vec<String>>>,
     ) -> impl Future<Output = Result<Arc<Language>>> {
         let filename = path.file_name().and_then(|name| name.to_str());
-        let extension = path.extension_or_hidden_file_name();
-        let path_suffixes = [extension, filename];
         let empty = Vec::new();
 
         let rx = self.get_or_load_language(move |language_name, config| {
-            let path_matches_default_suffix = config
-                .path_suffixes
-                .iter()
-                .any(|suffix| path_suffixes.contains(&Some(suffix.as_str())));
+            let filename = match filename {
+                Some(filename) => filename,
+                None => return 0,
+            };
+            let matching_default_suffix = config.path_suffixes.iter().fold(None, |acc, suffix| {
+                if filename.ends_with(suffix) {
+                    match acc {
+                        None => Some(suffix),
+                        Some(s) if suffix.len() > s.len() => Some(suffix),
+                        Some(_) => acc,
+                    }
+                } else {
+                    None
+                }
+            });
             let path_matches_custom_suffix = user_file_types
                 .and_then(|types| types.get(language_name))
                 .unwrap_or(&empty)
                 .iter()
-                .any(|suffix| path_suffixes.contains(&Some(suffix.as_str())));
+                .any(|suffix| filename.ends_with(suffix));
```

Comment:
Oops, I forgot to update this like I did above to support the full file name and prevent (eg) `rs` also matching `.cars`. I'll leave this as is for now, pending input from someone else on how to proceed. 😄

Reply:
Yeah, I think we should use the same logic for finding the suffix length, for both the default path suffixes and the user-defined path suffixes.

```rs
let matching_suffix_len = config
    .path_suffixes
    .iter()
    .filter(|suffix| {
        filename == suffix ||
            filename.strip_suffix(suffix).map_or(false, |prefix| prefix.ends_with('.'))
    })
    .map(|suffix| suffix.len())
    .max();
```